ouroboros-consensus/test/storage-test/Test/Ouroboros/Storage/LedgerDB/V1/BackingStore/Lockstep.hs

```diff
@@ -0,0 +1,811 @@
+  where
+    wrap ::
+         (a -> BSVal ks vs d b)
+      -> (Either Err a, Mock vs)
+      -> (BSVal ks vs d (Either Err b), Mock vs)
+    wrap f = first (MEither . bimap MErr f)
+
+    getHandle :: BSVal ks vs d Handle -> ValueHandle vs
+    getHandle (MValueHandle h) = h
+
+{-------------------------------------------------------------------------------
+  Generator
+-------------------------------------------------------------------------------}
+
+arbitraryBackingStoreAction ::
+     forall ks vs d.
+     ( Eq ks, Eq vs, Eq d, Typeable vs
+     , QC.Arbitrary ks, QC.Arbitrary vs
+     , QC.Arbitrary (BS.RangeQuery ks)
+     , Mock.MakeDiff vs d
+     )
+  => ModelFindVariables (BackingStoreState ks vs d)
+  -> BackingStoreState ks vs d
+  -> Gen (Any (LockstepAction (BackingStoreState ks vs d)))
+arbitraryBackingStoreAction findVars (BackingStoreState mock _stats) =
+    QC.frequency $
+         withoutVars
+      ++ case findVars (Proxy @(Either Err Handle)) of
+          []   -> []
+          vars -> withVars (QC.elements vars)
+  where
+    withoutVars :: [(Int, Gen (Any (LockstepAction (BackingStoreState ks vs d))))]
+    withoutVars = [
+        (5, fmap Some $ BSInitFromValues <$> QC.arbitrary <*> (Values <$> QC.arbitrary))
+      , (5, fmap Some $ BSInitFromCopy <$> genBackingStorePath)
+      , (2, pure $ Some BSClose)
+      , (5, fmap Some $ BSCopy <$> genBackingStorePath)
+      , (5, pure $ Some BSValueHandle)
+      , (5, fmap Some $ BSWrite <$> genSlotNo <*> genDiff)
+      ]
+
+    withVars ::
+         Gen (BSVar ks vs d (Either Err Handle))
+      -> [(Int, Gen (Any (LockstepAction (BackingStoreState ks vs d))))]
+    withVars genVar = [
+          (5, fmap Some $ BSVHClose <$> (fhandle <$> genVar))
+        , (5, fmap Some $ BSVHRangeRead <$> (fhandle <$> genVar) <*> QC.arbitrary)
+        , (5, fmap Some $ BSVHRead <$> (fhandle <$> genVar) <*> QC.arbitrary)
+        , (5, fmap Some $ BSVHAtSlot <$> (fhandle <$> genVar))
+        , (5, fmap Some $ BSVHStat <$> (fhandle <$> genVar))
+        ]
+      where
+        fhandle ::
+             GVar Op (Either Err Handle)
+          -> GVar Op Handle
+        fhandle = mapGVar (\op -> OpRight `OpComp` op)
+
+    genBackingStorePath :: Gen FS.FsPath
+    genBackingStorePath = do
+      file <- genBSPFile
+      pure . mkFsPath $ ["copies", file]
+
+    -- Generate a file name for a copy of the backing store contents. We keep
+    -- the set of possible file names small, such that errors (i.e., file alread
+    -- exists) occur most of the time.
+    genBSPFile :: Gen String
+    genBSPFile = QC.elements [show x | x <- [1 :: Int .. 10]]
+
+    -- Generate a slot number that is close before, at, or after the backing
+    -- store's current slot number. A
+    genSlotNo :: Gen SlotNo
+    genSlotNo = do
+        n :: Int <- QC.choose (-5, 5)
+        pure $ maybe 0 (+ fromIntegral n) (withOriginToMaybe seqNo)
```

Comment:
Are you intentionally allowing for negative numbers here?

Reply:
Yeah, this intentionally tests an unhappy path. The SUT will throw an error and the model checks that it is the error we are expecting

However, now that I'm looking at it, we might make the happy path a little more likely, by picking the positive number a bit higher

Reply:
Done